File: CSharpProject/CSharpProject/Node.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharpProject
{
    partial class AhoKorasik
    {
        private class Node
        {
            public Dictionary<char, Node> Sons;
            public Dictionary<char, Node> Transitions;
            public bool IsTerminal;
            public Node SuffLink;
            public Node PressedSuffixLink;
            public Node Parent;
            public char CharToParent;

            public Node(Node parent = null, char charToParent = (char)0)
            {
                Sons = new Dictionary<char, Node>();
                Transitions = new Dictionary<char, Node>();
                IsTerminal = false;
                SuffLink = null;
                Parent = parent;
                CharToParent = charToParent;
            }

            public Node GetTransition(char key)
            {
                if (!Transitions.ContainsKey(key))
                {
                    if (Sons.ContainsKey(key))
                    {
                        Transitions[key] = Sons[key];
                    }
                    else if (Parent == null)
                    {
                        Transitions[key] = this;
                    }
                    else
                    {
                        Transitions[key] = GetSuffLink().GetTransition(key);
                    }
                }
                return Transitions[key];
            }

            public Node GetSuffLink()
            {
                if (SuffLink == null)
                {
                    if (Parent == null)
                    {
                        SuffLink = this;
                    }
                    else if (Parent.Parent == null)
                    {
                        SuffLink = Parent;
                    }
                    else
                    {
                        SuffLink = Parent.GetSuffLink().GetTransition(CharToParent);
                    }
                }
                return SuffLink;
            }

            public Node GetPressedSuffixLink()
            {
                if (PressedSuffixLink == null)
                {
                    if (GetSuffLink().IsTerminal || GetSuffLink().Parent == null)
                    {
                        PressedSuffixLink = GetSuffLink();
                    }
                    else
                    {
                        PressedSuffixLink = GetSuffLink().GetPressedSuffixLink();
                    }
                }
                return PressedSuffixLink;
            }
        }
    }
}

File: CPlusPlusProject/CPlusPlusProject/Node.cpp
#include "Node.h"
#include <map>
#include <omp.h>

Node::Node(Node* parent, char charToParent)
{
    Sons = std::map<char, Node*>();
    Transitions = std::map<char, Node*>();
    IsTerminal = false;
    SuffLink = NULL;
    PressedSuffixLink = NULL;
    Parent = parent;
    CharToParent = charToParent;

    omp_init_lock(&Locker);
}

Node::~Node()
{
    for (auto it = Sons.begin(); it != Sons.end(); ++it)
    {
        delete it->second;
    }

    Sons.clear();
    Transitions.clear();
}

Node* Node::NonLockedGetTransition(char key)
{
    if (Transitions.find(key) == Transitions.end())
    {
        if (Sons.find(key) != Sons.end())
        {
            Transitions[key] = Sons[key];
        }
        else if (Parent == NULL)
        {
            Transitions[key] = this;
        }
        else
        {
            Transitions[key] = NonLockedGetSuffLink()->GetTransition(key);
        }
    }
    return Transitions[key];
}

Node* Node::NonLockedGetSuffLink()
{
    if (SuffLink == NULL)
    {
        if (Parent == NULL)
        {
            SuffLink = this;
        }
        else if (Parent->Parent == NULL)
        {
            SuffLink = Parent;
        }
        else
        {
            SuffLink = Parent->GetSuffLink()->GetTransition(CharToParent);
        }
    }
    return SuffLink;
}

Node* Node::NonLockedGetPressedSuffixLink()
{
    if (PressedSuffixLink == NULL)
    {
        if (NonLockedGetSuffLink()->IsTerminal || NonLockedGetSuffLink()->Parent == NULL)
        {
            PressedSuffixLink = NonLockedGetSuffLink();
        }
        else
        {
            PressedSuffixLink = NonLockedGetSuffLink()->GetPressedSuffixLink();
        }
    }
    return PressedSuffixLink;
}

Node* Node::GetTransition(char key)
{
    omp_set_lock(&Locker);

    Node * result = NonLockedGetTransition(key);

    omp_unset_lock(&Locker);
    return result;
}

Node* Node::GetSuffLink()
{
    omp_set_lock(&Locker);

    Node * result = NonLockedGetSuffLink();

    omp_unset_lock(&Locker);
    return result;
}

Node* Node::GetPressedSuffixLink()
{
    omp_set_lock(&Locker);

    Node * result = NonLockedGetPressedSuffixLink();

    omp_unset_lock(&Locker);
    return result;
}